### PR TITLE
Improving Index to Log Comparison in DataDump Tool

### PR DESCRIPTION
**Pull request by nsivabalan**
Improving the way we find holes in index to log compare operation in DumpData.

SLA: 15 mins
Reviewers: Gopal, Casey

Example: (no holes were found in the replica I tested against)
Index to log comparison output:
2016-11-09 12:35:06 INFO DumpData:214 - Type of Operation CompareIndexToLog
2016-11-09 12:35:06 INFO DumpData:216 - File to read /Users/sinaraya/logContentstemp/11229049741_index
2016-11-09 12:35:06 INFO DumpData:613 - Comparing Index entries to Log
2016-11-09 12:35:06 INFO DumpData:639 - Dumping index 11229049741_index for null

All indexes to log comparison output:
2016-11-09 12:45:06 INFO DumpData:209 - Type of Operation CompareReplicaIndexesToLog
2016-11-09 12:45:06 INFO DumpData:560 - Comparing Index entries to Log
2016-11-09 12:45:06 INFO DumpData:631 - Dumping index 0_index for logContentstemp
2016-11-09 12:45:13 INFO DumpData:631 - Dumping index 10612526754_index for logContentstemp
2016-11-09 12:45:28 INFO DumpData:631 - Dumping index 11229049741_index for logContentstemp
2016-11-09 12:45:44 INFO DumpData:631 - Dumping index 11848704171_index for logContentstemp
2016-11-09 12:45:58 INFO DumpData:631 - Dumping index 12476544102_index for logContentstemp
2016-11-09 12:46:11 INFO DumpData:631 - Dumping index 13139107375_index for logContentstemp
2016-11-09 12:46:24 INFO DumpData:631 - Dumping index 13798374621_index for logContentstemp
2016-11-09 12:46:37 INFO DumpData:631 - Dumping index 1437329806_index for logContentstemp
2016-11-09 12:46:51 INFO DumpData:631 - Dumping index 14438506657_index for logContentstemp
2016-11-09 12:47:04 INFO DumpData:631 - Dumping index 15079180828_index for logContentstemp
2016-11-09 12:47:16 INFO DumpData:631 - Dumping index 15741151234_index for logContentstemp
2016-11-09 12:47:29 INFO DumpData:631 - Dumping index 16387805631_index for logContentstemp
2016-11-09 12:47:42 INFO DumpData:631 - Dumping index 17017994734_index for logContentstemp
2016-11-09 12:48:00 INFO DumpData:631 - Dumping index 17668061472_index for logContentstemp
2016-11-09 12:48:17 INFO DumpData:631 - Dumping index 18391273296_index for logContentstemp
2016-11-09 12:48:32 INFO DumpData:631 - Dumping index 19045110343_index for logContentstemp
2016-11-09 12:48:44 INFO DumpData:631 - Dumping index 19670605237_index for logContentstemp
2016-11-09 12:48:58 INFO DumpData:631 - Dumping index 20283341996_index for logContentstemp
2016-11-09 12:48:59 INFO DumpData:631 - Dumping index 2472690072_index for logContentstemp
2016-11-09 12:49:17 INFO DumpData:631 - Dumping index 362344493_index for logContentstemp
2016-11-09 12:49:22 INFO DumpData:631 - Dumping index 3948801476_index for logContentstemp
2016-11-09 12:49:34 INFO DumpData:631 - Dumping index 4505100878_index for logContentstemp
2016-11-09 12:49:45 INFO DumpData:631 - Dumping index 484743025_index for logContentstemp
2016-11-09 12:49:51 INFO DumpData:631 - Dumping index 4995398370_index for logContentstemp
2016-11-09 12:50:03 INFO DumpData:631 - Dumping index 5496525249_index for logContentstemp
2016-11-09 12:50:15 INFO DumpData:631 - Dumping index 5997887805_index for logContentstemp
2016-11-09 12:50:29 INFO DumpData:631 - Dumping index 6539196504_index for logContentstemp
2016-11-09 12:50:33 ERROR DumpDataHelper:405 - Expiration value mismatch for AAEAAQAAAAAAAAAAAAAAJDQ5NWY0YmQ0LTkzNWEtNDk5Mi04NmE3LTI0NDI4NzlkMDc1YQ Index value false, Log value true, index TTL in ms 2747866697988, log Time to live in secs 1282617439, in ms 1282617439000
2016-11-09 12:50:41 INFO DumpData:631 - Dumping index 7078553003_index for logContentstemp
2016-11-09 12:50:57 INFO DumpData:631 - Dumping index 735261636_index for logContentstemp
2016-11-09 12:51:09 INFO DumpData:631 - Dumping index 7616198781_index for logContentstemp
2016-11-09 12:51:23 INFO DumpData:631 - Dumping index 8181826373_index for logContentstemp
2016-11-09 12:51:35 INFO DumpData:631 - Dumping index 8803462152_index for logContentstemp
2016-11-09 12:51:48 INFO DumpData:631 - Dumping index 9374186926_index for logContentstemp
2016-11-09 12:52:01 INFO DumpData:631 - Dumping index 9983801246_index for logContentstemp

### CODE REVIEW

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 1]
```diff
@@ -667,20 +657,41 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
+          if (isDeleted && deletedMsgOriginalOffsets != null) {
+            long originalOffset = blobValue.getOriginalMessageOffset();
+            deletedMsgOriginalOffsets.add(originalOffset);
+            if (originalOffset != -1 && !coveredRanges.containsKey(originalOffset) &&
+                originalOffset > startOffset) {
+              randomAccessFile.seek(originalOffset);
+              DumpDataHelper.BlobRecordInfo blobRecordInfo =
+                  dumpDataHelper.readSingleRecordFromLog(randomAccessFile, originalOffset);
+              coveredRanges.put(originalOffset, originalOffset + blobRecordInfo.totalRecordSize);
+              logger.trace("PUT Record " + blobRecordInfo.blobId + " added with orig offset " + originalOffset
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: use format string.

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 2]
```diff
@@ -667,20 +657,41 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
+                  + " and end offset " + (originalOffset + blobRecordInfo.totalRecordSize) + " for a delete msg "
+                  + key.getID() + " at offset " + blobValue.getOffset());
+              if (!blobRecordInfo.blobId.getID().equals(key.getID())) {
+                logger.error("BlobId value mismatch between delete record " + key.getID() + " and put record "
+                    + blobRecordInfo.blobId.getID());
+              }
+            } else {
+              if (coveredRanges.containsKey(originalOffset)) {
+                logger.error("Missed out Put Record ignored at " + originalOffset + " with delete msg's offset "
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: "missed out put record"? Didn't quite understand this log message.

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 3]
```diff
@@ -667,20 +657,41 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
-                  coveredRangesInLog);
+                  coveredRanges);
           if (!success) {
             logger.error("Failed for Index Entry " + msg);
           }
+          if (isDeleted && deletedMsgOriginalOffsets != null) {
+            long originalOffset = blobValue.getOriginalMessageOffset();
+            deletedMsgOriginalOffsets.add(originalOffset);
+            if (originalOffset != -1 && !coveredRanges.containsKey(originalOffset) &&
+                originalOffset > startOffset) {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: `originalOffset >= startOffset`?

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 4]
```diff
@@ -667,20 +657,41 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
+                  + key.getID() + " at offset " + blobValue.getOffset());
+              if (!blobRecordInfo.blobId.getID().equals(key.getID())) {
+                logger.error("BlobId value mismatch between delete record " + key.getID() + " and put record "
+                    + blobRecordInfo.blobId.getID());
+              }
+            } else {
+              if (coveredRanges.containsKey(originalOffset)) {
+                logger.error("Missed out Put Record ignored at " + originalOffset + " with delete msg's offset "
+                    + blobValue.getOffset() + " for an already available put record ");
+              } else if (!coveredRanges.containsKey(originalOffset) && startOffset > originalOffset) {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: looks like this can just be `else`?

File: ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java [hunk 5]
```diff
@@ -255,7 +255,7 @@ public BlobRecordInfo(String messageHeader, BlobId blobId, String blobProperty,
    * @throws IOException
    * @throws MessageFormatException
    */
-  private BlobRecordInfo readSingleRecordFromLog(RandomAccessFile randomAccessFile, long currentOffset)
+  public BlobRecordInfo readSingleRecordFromLog(RandomAccessFile randomAccessFile, long currentOffset)
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java]: why is this `public` ? the class itself is package-private so this being `public` doesn't seem right.

File: ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java [hunk 6]
```diff
@@ -219,7 +220,7 @@ public void dumpLog(File file, long startOffset, long endOffset, ArrayList<Strin
   /**
    * Holds information about a blob record in the log
    */
-  private class BlobRecordInfo {
+  class BlobRecordInfo {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java]: any reason to make this package private?

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 7]
```diff
@@ -562,40 +559,31 @@ public void compareReplicaIndexEntriestoLogContent(String replicaRootDirectory,
...
-    MergedIntervals coveredRangesInLog = new MergedIntervals();
-    Map<Long, Long> deletedMsgInfo = new HashMap<>();
     for (File indexFile : replicaDirectory.listFiles(new IndexFileNameFilter())) {
-      compareIndexEntriesToLogContent(indexFile, replicaDirectory, randomAccessFile, coveredRangesInLog, deletedMsgInfo,
-          false);
+      TreeMap<Long, Long> coveredRanges = new TreeMap<Long, Long>();
+      Set<Long> deleteMsgsOriginalOffsets = new HashSet<Long>();
+      compareIndexEntriesToLogContent(indexFile, replicaDirectory, randomAccessFile, coveredRanges,
+          deleteMsgsOriginalOffsets);
     }
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: We can do our best in trying to get the first offset of first segment and end offset of last segment. But still we might have to parse the logs to know the holes at the end. But to the minimum, we can do what index consistency checker does if you feel strongly.

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 8]
```diff
@@ -562,40 +559,31 @@ public void compareReplicaIndexEntriestoLogContent(String replicaRootDirectory,
...
+      Map.Entry<Long, Long> curEntry = iterator.next();
+      logger.trace("Record startOffset " + curEntry.getKey() + ", endOffset " + curEntry.getValue());
+      if (prevEntry.getValue().compareTo(curEntry.getKey()) != 0) {
+        logger.error(
+            "Cannot find entries in Index ranging from " + prevEntry.getValue() + " to " + curEntry.getKey()
+                + " with a hole of " + (curEntry.getKey() - prevEntry.getValue()) + " in the Log");
       }
-      previousRange = curRange;
+      prevEntry = curEntry;
     }
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: will add it

File: config/DataDumplog4j.props [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# Root logger option
```
Comment (nsivabalan) [inline config/DataDumplog4j.props]: This has options to either redirect to stdout or to a file. Also, might add more options later for tools. So want to have a separate one.

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java (hunk 7 above)
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: The "holes at the end" problem is ok (we can find a way around it) - but I think checking that all the index segments together cover the whole log is a requirement.

File: ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java [hunk 10]
```diff
@@ -347,17 +348,17 @@ private long getLogOffsetEnd(File indexFile)
       }
       return fileEndPointer;
     }
+  }
 
-    class IndexFileNameComparator implements Comparator {
-      public int compare(Object o1, Object o2) {
-        String s1 = ((File) o1).getName();
-        String s2 = ((File) o2).getName();
+  static class IndexFileNameComparator implements Comparator {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java]: are you going to do all the "reuse store code" changes in another commit? Because all of this functionality is already available and this is repeated code.

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 11]
```diff
@@ -667,20 +651,39 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
+              coveredRanges.put(originalOffset, originalOffset + blobRecordInfo.totalRecordSize);
+              logger.trace(
+                  "PUT Record {} added with orig offset {} and end offset {} for a delete msg " + "{} at offset {} ",
+                  blobRecordInfo.blobId, originalOffset, (originalOffset + blobRecordInfo.totalRecordSize), key.getID(),
+                  blobValue.getOffset());
+              if (!blobRecordInfo.blobId.getID().equals(key.getID())) {
+                logger.error("BlobId value mismatch between delete record " + key.getID() + " and put record "
+                    + blobRecordInfo.blobId.getID());
+              }
+            } else {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: why not an `else if` instead of
```
else {
  if {
  }
}
```

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 12]
```diff
@@ -667,20 +651,39 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
               dumpDataHelper.readFromLogAndVerify(randomAccessFile, blobValue.getOffset(), key.getID(), blobValue,
-                  coveredRangesInLog);
+                  coveredRanges);
           if (!success) {
             logger.error("Failed for Index Entry " + msg);
           }
+          if (isDeleted) {
+            long originalOffset = blobValue.getOriginalMessageOffset();
+            deleteMsgsOriginalOffsets.add(originalOffset);
+            if (originalOffset != -1 && !coveredRanges.containsKey(originalOffset) &&
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: do you need the `originalOffset != 1` condition? Won't the `originalOffset >= startOffset` take care of this ?

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 13]
```diff
@@ -667,20 +651,39 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
+              logger.trace(
+                  "PUT Record {} added with orig offset {} and end offset {} for a delete msg " + "{} at offset {} ",
+                  blobRecordInfo.blobId, originalOffset, (originalOffset + blobRecordInfo.totalRecordSize), key.getID(),
+                  blobValue.getOffset());
+              if (!blobRecordInfo.blobId.getID().equals(key.getID())) {
+                logger.error("BlobId value mismatch between delete record " + key.getID() + " and put record "
+                    + blobRecordInfo.blobId.getID());
+              }
+            } else {
+              if (!coveredRanges.containsKey(originalOffset) && startOffset > originalOffset) {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: Not sure I understand this check - if the second condition is `true`, how the can the first condition ever have been `true` - considering how your previous `if` is structured
```
if (originalOffset != -1 && !coveredRanges.containsKey(originalOffset) && originalOffset >= startOffset)
```
You only ever add to `coveredRanges` if `originalOffset >= startOffset`. I might be missing something but the checks seem extraneous - let's keep it minimum and simple if possible

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 14]
```diff
@@ -667,20 +651,39 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
+                logger.trace("Put Record ignored at {} with delete msg's offset {} prior to startOffset {} ",
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: log message is a bit misleading. Seems to say that the deleted message's offset is prior to start offset (which it is not). 

Should be something like "Put Record at {} with delete msg offset {} ignored because it is prior to startOffset {}"

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java [hunk 15]
```diff
@@ -667,20 +651,39 @@ private void compareIndexEntriesToLogContent(File indexFile, File replicaDirecto
...
-            deletedMsgInfo.put(blobValue.getOriginalMessageOffset(), blobValue.getOffset());
-          }
           boolean success =
               dumpDataHelper.readFromLogAndVerify(randomAccessFile, blobValue.getOffset(), key.getID(), blobValue,
-                  coveredRangesInLog);
+                  coveredRanges);
           if (!success) {
             logger.error("Failed for Index Entry " + msg);
           }
+          if (isDeleted) {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: why continue with this record if `success` is `false`? To account for it's PUT record?

File: ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java (hunk 10 above)
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java]: yes. Will make another commit

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java (hunk 15 above)
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: Good point. I didn't consider that

File: ambry-tools/src/main/java/com.github.ambry/store/DumpData.java (hunk 13 above)
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpData.java]: have fixed it

**Comment by nsivabalan** [comment]
@vgkholla : Have addressed your comments.

File: ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java [hunk 16]
```diff
@@ -140,44 +144,52 @@ public void dumpLog(File file, long startOffset, long endOffset, ArrayList<Strin
       long tempCurrentOffset = currentOffset;
 
       try {
-        BlobRecordInfo blobRecordInfo = readSingleRecordFromLog(randomAccessFile, currentOffset);
+        LogBlobRecordInfo logBlobRecordInfo = readSingleRecordFromLog(randomAccessFile, currentOffset);
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java]: where is `LogBlobRecordInfo`? Doesn't seem to be on this PR or on the master branch.

Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java]: Have added it now for real.

**Comment by vgkholla** [comment]
LGTM. Please confirm build and formatting

**Comment by nsivabalan** [comment]
yes. done.

File: ambry-tools/src/main/java/com.github.ambry/store/LogBlobRecordInfo.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.commons.BlobId;
+
+/**
+ * Holds information about a blob record in the log
+ */
+public class LogBlobRecordInfo {
+  String messageHeader = null;
```
Comment (cgtz) [inline ambry-tools/src/main/java/com.github.ambry/store/LogBlobRecordInfo.java]: Could these be final too?

File: ambry-tools/src/main/java/com.github.ambry/store/LogBlobRecordInfo.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.commons.BlobId;
+
+/**
+ * Holds information about a blob record in the log
+ */
+public class LogBlobRecordInfo {
```
Comment (cgtz) [inline ambry-tools/src/main/java/com.github.ambry/store/LogBlobRecordInfo.java]: Could this be package-private?

File: ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java [hunk 19]
```diff
@@ -49,12 +48,10 @@
 public class DumpDataHelper {
```
Comment (cgtz) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java]: Could this class be package-private too?

File: ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java [hunk 20]
```diff
@@ -353,35 +333,34 @@ public void dumpReplicaToken(File replicaTokenFile) throws Exception {
    * @throws IOException
```
Comment (cgtz) [inline ambry-tools/src/main/java/com.github.ambry/store/DumpDataHelper.java]: add info about `coveredRanges` to this javadoc

File: ambry-tools/src/main/java/com.github.ambry/store/BlobStatus.java [hunk 21]
```diff
@@ -14,18 +14,20 @@
...
 import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
 
 
 public class BlobStatus {
-  ArrayList<String> available;
-  ArrayList<String> deletedOrExpired;
-  ArrayList<String> unavailable;
+  Set<String> available;
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/store/BlobStatus.java]: can't be made final as we update as and when required

**Comment by nsivabalan** [comment]
@cgtz : Addressed your comments.
